core/src/main/scala/scalaz/StrictTree.scala

```diff
@@ -0,0 +1,449 @@
+
+  def foldNode[Z](f: A => Vector[StrictTree[A]] => Z): Z =
+    f(rootLabel)(subForest)
+
+  /**
+    * This implementation is 10x faster than mapTrampoline for StrictTreeTestJVM's map test.
+    */
+  private def mapCombiner[B](f: A => B)(rootLabel: A)(subForest: Seq[StrictTree[B]]): StrictTree[B] = {
+    StrictTree[B](f(rootLabel), subForest.toVector)
+  }
+
+  def map[B](f: A => B): StrictTree[B] = {
+    runBottomUp(mapCombiner(f))
+  }
+
+  /**
+    * This implementation is 9x faster than flatMapTrampoline for StrictTreeTestJVM's flatMap test.
+    */
+  private def flatMapCombiner[B](f: A => StrictTree[B])(root: A)(subForest: Seq[StrictTree[B]]): StrictTree[B] = {
+    val StrictTree(rootLabel0, subForest0) = f(root)
+    StrictTree(rootLabel0, subForest0 ++ subForest)
+  }
+
+  def flatMap[B](f: A => StrictTree[B]): StrictTree[B] = {
+    runBottomUp(flatMapCombiner(f))
+  }
+
+  def traverse1[G[_] : Apply, B](f: A => G[B]): G[StrictTree[B]] = {
+    val G = Apply[G]
+
+    subForest match {
+      case Vector() => G.map(f(rootLabel))(Leaf(_))
+      case x +: xs => G.apply2(f(rootLabel), NonEmptyList.nel(x, IList.fromFoldable(xs)).traverse1(_.traverse1(f))) {
+        case (h, t) => Node(h, t.list.toVector)
+      }
+    }
+  }
+
+  private case class ZipStackElem[B](
+    parent: Option[ZipStackElem[B]],
+    a: StrictTree[A],
+    b: StrictTree[B]
+  ) extends Iterator[(StrictTree[A], StrictTree[B])] {
+    private val zippedSubIterator =
+      a.subForest.iterator.zip(b.subForest.iterator)
+
+    val mappedSubForest: mutable.Buffer[StrictTree[(A, B)]] = mutable.Buffer.empty
+
+    override def hasNext: Boolean = zippedSubIterator.hasNext
+
+    override def next(): (StrictTree[A], StrictTree[B]) = zippedSubIterator.next()
+  }
+
+  def zip[B](b: StrictTree[B]): StrictTree[(A, B)] = {
+    val root = ZipStackElem[B](None, this, b)
+    val stack = mutable.Stack[ZipStackElem[B]](root)
+
+    while (stack.nonEmpty) {
+      val here = stack.elems.head
+      if (here.hasNext) {
+        val (childA, childB) = here.next()
+        val nextStackElem = ZipStackElem[B](Some(here), childA, childB)
+        stack.push(nextStackElem)
+      } else {
+        //The "here" node is completed, so add its result to its parents completed children.
+        val result = StrictTree((here.a.rootLabel, here.b.rootLabel), here.mappedSubForest.toVector)
+        here.parent.foreach(_.mappedSubForest += result)
+        stack.pop()
+      }
+    }
+
+    StrictTree((rootLabel, b.rootLabel), root.mappedSubForest.toVector)
+  }
+
+  private def hashCodeCombiner(root: A)(subForest: Seq[Int]): Int = {
+    root.hashCode ^ subForest.hashCode
+  }
+
+  /**
+    * This implementation is 24x faster than the trampolined implementation for StrictTreeTestJVM's hashCode test.
+    *
+    * @return
+    */
+  override def hashCode(): Int = {
+    runBottomUp(hashCodeCombiner)
+  }
+
+  override def equals(obj: scala.Any): Boolean = {
```

Comment:
The reason for this override is to avoid stack overflow on recursive calls to equals.